waiter/src/waiter/service_description.clj

```diff
@@ -276,38 +276,56 @@
                                              "version" "default-version"
                                              "run-as-user" "default-run-as-user"})
         service-description-to-use (merge default-valid-service-description service-description-template)
-        exception-message (utils/message :invalid-service-description)]
+        exception-message (utils/message :invalid-service-description)
+        throw-error (fn throw-error [e issue friendly-error-message]
+                      (sling/throw+ (cond-> {:type :service-description-error
+                                             :message exception-message
+                                             :service-description service-description-template
+                                             :status 400
+                                             :issue issue}
+                                            (not-empty friendly-error-message) (assoc :friendly-error-message friendly-error-message))
+                                    e))]
     (try
       (s/validate service-description-schema service-description-to-use)
       (catch Exception e
         (let [issue (s/check service-description-schema service-description-to-use)
               friendly-error-message (utils/filterm val
                                                     {:metadata (generate-friendly-metadata-error-message issue)
                                                      :env (generate-friendly-environment-variable-error-message issue)})]
-          (sling/throw+ (cond-> {:type :service-description-error
-                                 :message exception-message
-                                 :service-description service-description-template
-                                 :status 400
-                                 :issue issue}
-                                (not-empty friendly-error-message) (assoc :friendly-error-message friendly-error-message))
-                        e))))
+          (throw-error e issue friendly-error-message))))
+
+    (try
+      (s/validate upper-limits-schema service-description-to-use)
+      (catch Exception e
+        (let [issue (s/check upper-limits-schema service-description-to-use)
+              issue->param->limit (fn [issue param]
+                                    (-> issue (get param) .schema .pred-name (str/replace "limit-" "")))
+              param->message (fn [param]
+                               (str "- " param " is " (get service-description-to-use param) " but the max allowed is "
```

Comment:
The use of newlines assumes the user is expecting `text/plain`, which, when rendered in HTML, will look funny.

Reply:
Removed the newlines. 

@dposada do you agree with @sradack?